File: ios/Objective-C/GeneralSettingsObjC/GeneralSettingsObjC/Modules/BasicCell/BasicPullTableViewCell.m
//
//  BasicPullTableViewCell.m
//  GeneralSettings
//
//  Created by dynamsoft on 2021/11/26.
//

#import "BasicPullTableViewCell.h"

@implementation BasicPullTableViewCell

+ (CGFloat)cellHeight
{
    return kCellHeight;
}

- (instancetype)initWithStyle:(UITableViewCellStyle)style reuseIdentifier:(NSString *)reuseIdentifier
{
    self = [super initWithStyle:style reuseIdentifier:reuseIdentifier];
    if (self) {
        [self setupUI];
    }
    return self;
}

- (void)setupUI
{
    self.selectionStyle = UITableViewCellSelectionStyleNone;
    
    self.titleLabel.frame = CGRectMake(kCellLeftMargin, 0, 0, kCellHeight);
    self.titleLabel.font = kFont_Regular(kCellTitleFontSize);
    self.titleLabel.textAlignment = NSTextAlignmentLeft;
    [self.contentView addSubview:self.titleLabel];
    
    self.pullImageV.frame = CGRectMake(kScreenWidth - kCellRightMargin - 16 * kScreenAdaptationRadio, (kCellHeight - 16 * kScreenAdaptationRadio) / 2.0, 16 * kScreenAdaptationRadio, 16 * kScreenAdaptationRadio);
    self.pullImageV.image = [UIImage imageNamed:@"select_down"];
    [self.contentView addSubview:self.pullImageV];
    
    self.pullTextLabel.frame = CGRectMake(self.pullImageV.left - 150 * kScreenAdaptationRadio, (kCellHeight - 20 * kScreenAdaptationRadio) / 2.0, 150 * kScreenAdaptationRadio, 20 * kScreenAdaptationRadio);
    self.pullTextLabel.font = kFont_Regular(kCellPullTextFontSize);
    self.pullTextLabel.textAlignment = NSTextAlignmentRight;
    self.pullTextLabel.text = @"";
    [self.contentView addSubview:self.pullTextLabel];
    
    self.separationLine.frame = CGRectMake(0, kCellHeight - KCellSeparationLineHeight, kScreenWidth, KCellSeparationLineHeight);
    self.separationLine.backgroundColor = kCellSeparationLineBackgroundColor;
    [self.contentView addSubview:self.separationLine];
}

/// Update UI.
- (void)updateUIWithTitle:(NSString *)titleString andContentString:(NSString *)contentString
{
    self.titleLabel.text = titleString;
    self.titleLabel.width = [[ToolsHandle toolManger] calculateWidthWithText:titleString font:self.titleLabel.font AndComponentheight:self.titleLabel.height];
    
    self.pullTextLabel.text = contentString;
    
}

// MARK: - Lazy loading

- (UILabel *)titleLabel
{
    if (!_titleLabel) {
        _titleLabel = [[UILabel alloc] init];
    }
    return _titleLabel;
}

- (UIImageView *)pullImageV
{
    if (!_pullImageV) {
        _pullImageV = [[UIImageView alloc] init];
    }
    return _pullImageV;
}

- (UILabel *)pullTextLabel
{
    if (!_pullTextLabel) {
        _pullTextLabel = [[UILabel alloc] init];
    }
    return _pullTextLabel;
}

- (UIView *)separationLine
{
    if (!_separationLine) {
        _separationLine = [[UIView alloc] init];
    }
    return _separationLine;
}

@end

File: ios/Swift/GeneralSettingsSwift/GeneralSettingsSwift/ViewControllers/Cells/BasicPullTableViewCell.swift
//
//  BasicPullTableViewCell.swift
//  GeneralSettingsSwift
//
//  Copyright © Dynamsoft. All rights reserved.
//

import UIKit

class BasicPullTableViewCell: UITableViewCell {

    private lazy var titleLabel: UILabel = {
        let label = UILabel.init(frame: CGRectMake(kCellLeftMargin, 0, 0, kCellHeight))
        label.font = kFont_Regular(kCellTitleFontSize)
        label.textAlignment = .left
        return label
    }()
    
    private lazy var pullImageV: UIImageView = {
        let image = UIImageView.init(frame: CGRectMake(kScreenWidth - kCellRightMargin - 16, (kCellHeight - 16) / 2.0, 16, 16))
        image.image = UIImage(named: "select_down")
        return image
    }()
    
    private lazy var pullTextLabel: UILabel = {
        let label = UILabel.init(frame: CGRectMake(self.pullImageV.left - 150, (kCellHeight - 20) / 2.0, 150, 20))
        label.font = kFont_Regular(kCellPullTextFontSize)
        label.textAlignment = .right
        return label
    }()
    
    private lazy var separationLine: UIView = {
        let separationLine = UIView.init(frame: CGRectMake(0, kCellHeight - KCellSeparationLineHeight, kScreenWidth, KCellSeparationLineHeight))
        separationLine.backgroundColor = kCellSeparationLineBackgroundColor
        return separationLine
    }()
    
    static func cellHeight() -> CGFloat {
        return kCellHeight
    }
    
    override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)
        setupUI()
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    private func setupUI() -> Void {
        self.selectionStyle = .none
        
        self.contentView.addSubview(titleLabel)
        self.contentView.addSubview(pullImageV)
        self.contentView.addSubview(pullTextLabel)
        self.contentView.addSubview(separationLine)
    }
    
    /// Update UI.
    func updateUI(with title: String, content: String) -> Void {
        self.titleLabel.text = title
        self.titleLabel.width = ToolsManager.shared.calculateWidth(with: title, font: self.titleLabel.font, componentHeight: self.titleLabel.height)
        self.pullTextLabel.text = content
    }
}
